syntax = "proto3";

package unified_perf_format;

message Stage{
    string name = 1;
    uint64 order_id = 2;
    uint64 start_time = 3;
    uint64 end_time = 4;
    map<string, string> metadata = 5;
};

message Instruction{
    string name = 1;
    uint32 thread_id = 2;
    uint64 global_seq_num = 3;
    repeated uint64 parent_seq_num = 4;
    map<string, string> metadata = 5;
    repeated Stage stages = 6;
};

enum InstType{
   UNKNOWN = 0;
   CALL=1;
   RET=2;
   POINT_SHOW = 3;
};

message Function{
    string name = 1;
    uint32 thread_id = 2;
    uint64 pc = 3;
    uint64 timestamp = 4;
    InstType inst_type = 5;
    map<string, string> metadata = 6;
};

message CntValue{
    uint64 timestamp = 1;
    double value = 2;
};

message Counter{
    string name = 1;
    string unit = 2;
    repeated CntValue values = 3;
    map<string, string> metadata = 4;
};

message BatchInstruction{
    repeated Instruction instructions = 1;
};

message BatchFunction{
    repeated Function functions = 1;
};

message BatchCounter{
    repeated Counter counters = 1;
};

message UnifiedPerfData{
    enum DataType{
        INSTRUCTIONS = 0;
        FUNCTIONS = 1;
        COUNTERS = 2;
    };
    DataType data_type = 1;
    string device_name = 2;
    oneof payload{
        BatchInstruction instructions = 3;
        BatchFunction functions = 4;
        BatchCounter counters = 5;
    };
}

// 容器消息：包含多个 UnifiedPerfData，用于在一个文件中存储多种类型的数据
message UnifiedPerfDataContainer{
    repeated UnifiedPerfData data_list = 1;
}